### PR TITLE
feat: Update the go feature server from Expedia code repo.

### PR DESCRIPTION
**Pull request by shuchu**
# What this PR does / why we need it:

Update the improved Go feature server code from ExpediaGroup/Feast to our code repo.

# Which issue(s) this PR fixes:

### CODE REVIEW

**Comment by shuchu** [comment]
cc: @EXPEbdodla

**Comment by shuchu** [comment]
[Update 10-22-2024] Now it can compile, build and build docker image. All configures in the Makefile under the "#Go SDK & embedded" are working, except the "install-go-ci-dependencies".  

The current "install-go-ci-dependencies" is obsoleted. We need to fix it when we setup the integration test for Go feature server. 

!!Warning, the current version my not function correctly. need more testing and investigating.

**Comment by shuchu** [comment]
[Update 10-31-2024] 

1.  "make test-go" works now
2.  DataDog related observability instrumentation code were commented out but not removed. I keep these code and plan to use OTEL to replace them.

File: go/internal/feast/registry/http.go [hunk 1]
```diff
@@ -0,0 +1,212 @@
+package registry
```
Comment (EXPEbdodla) [inline go/internal/feast/registry/http.go:1]: This needs to be switched to remote registry. We built http registry to overcome the password sharing for SQL Registry. HTTP Registry which is FastAPI based is not contributed to Feast. We are also planning to leverage Remote registry and move over from HTTP one.

Comment (shuchu) [reply go/internal/feast/registry/http.go:1]: Thank you for the info, @EXPEBdodla ! Let me see try to resolve this.

**Comment by shuchu** [comment]
[Update 11/04/2024] 

1. The HTTP Registry code is removed as it depends on an external HTTP server as @EXPEbdodla pointed out.
2.  The default Registry is File based one now.

File: go/internal/feast/featurestore.go [hunk 2]
```diff
@@ -45,18 +50,29 @@ func NewFeatureStore(config *registry.RepoConfig, callback transformation.Transf
...
 		return nil, err
 	}
-	registry.InitializeRegistry()
+	err = registry.InitializeRegistry()
+	if err != nil {
+		return nil, err
+	}
+	sanitizedProjectName := strings.Replace(config.Project, "_", "-", -1)
+	productName := os.Getenv("PRODUCT")
+	endpoint := fmt.Sprintf("%s-transformations.%s.svc.cluster.local:80", sanitizedProjectName, productName)
```
Comment (shuchu) [inline go/internal/feast/featurestore.go:67]: Is this an external Transformation server? I guess we need to remove this and related code? @EXPEbdodla

Comment (EXPEbdodla) [reply go/internal/feast/featurestore.go:67]: Initially transformations are running with GOPY bindings. That's not working at scale. So we implemented it similar to Java Feature Server.

Comment (shuchu) [reply go/internal/feast/featurestore.go:67]: Thank you, @EXPEbdodla  I guess I need to work on this part.

File: go/internal/feast/registry/http.go (hunk 1 above)
Comment (shuchu) [reply go/internal/feast/registry/http.go]: I choose an "easy" fix for this.

File: go/internal/feast/featurestore.go (hunk 2 above)
Comment (shuchu) [reply go/internal/feast/featurestore.go]: I chose an "easy" fix for this. :crying_cat_face:

File: go/internal/feast/featurestore.go [hunk 3]
```diff
@@ -297,6 +315,10 @@ func (fs *FeatureStore) readFromOnlineStore(ctx context.Context, entityRows []*p
 	requestedFeatureViewNames []string,
 	requestedFeatureNames []string,
 ) ([][]onlinestore.FeatureData, error) {
+	// Create a Datadog span from context
```
Comment (HaoXuAI) [inline go/internal/feast/featurestore.go:318]: I think these can be removed or updated as datadog is not part of feast

File: go/internal/feast/onlinestore/redisonlinestore.go [hunk 4]
```diff
@@ -89,23 +98,42 @@ func NewRedisOnlineStore(project string, config *registry.RepoConfig, onlineStor
...
 				}
 			} else {
-				return nil, errors.New(fmt.Sprintf("unable to parse a part of connection_string: %s. Must contain either ':' (addresses) or '=' (options", part))
+				return nil, fmt.Errorf("unable to parse a part of connection_string: %s. Must contain either ':' (addresses) or '=' (options", part)
 			}
 		}
 	}
 
-	if t == redisNode {
+	// Metrics are not showing up when the service name is set to DD_SERVICE
```
Comment (HaoXuAI) [inline go/internal/feast/onlinestore/redisonlinestore.go:109]: similar here

File: go/README.md [hunk 5]
```diff
@@ -1,109 +1,12 @@
-This directory contains the Go logic that's executed by the `EmbeddedOnlineFeatureServer` from Python.
```
Comment (HaoXuAI) [inline go/README.md:1]: Any reason we are removing the document here?

File: go/internal/feast/featurestore.go (hunk 2 above)
Comment (EXPEbdodla) [reply go/internal/feast/featurestore.go]: Yes. We should get the transformation server endpoint from feature_store.yaml.

File: go/internal/feast/featurestore.go (hunk 3 above)
Comment (EXPEbdodla) [reply go/internal/feast/featurestore.go:318]: Yea. Datadog needs to be replaced with OTEL.

File: go/README.md (hunk 5 above)
Comment (EXPEbdodla) [reply go/README.md:1]: That piece of code is not needed if we remove GOPY bindings.

File: go/internal/feast/featurestore.go (hunk 3 above)
Comment (shuchu) [reply go/internal/feast/featurestore.go:318]: yes, I will replace them with OTEL after this PR get merged.

**Comment by shuchu** [comment]
[Update 11/24/2024]

1.  use the feature_store.yaml file as the source of defining Python transformation service.  Example:
```
# feature_store.yaml file
feature_store:
   transformation_service_endpoint: "localhost:50051"
```

@EXPEbdodla @feast-dev/reviewers-and-approvers

File: go/internal/feast/featurestore.go (hunk 2 above)
Comment (shuchu) [reply go/internal/feast/featurestore.go]: updated. :)

File: Makefile [hunk 6]
```diff
@@ -536,3 +536,40 @@ build-helm-docs:
...
+
+# Go SDK & embedded
+install-protoc-dependencies:
+	pip install "protobuf>=4.24.0,<5.0.0" "grpcio-tools>=1.56.2,<2" "mypy-protobuf>=3.1"
+
+install-go-proto-dependencies:
+	go install google.golang.org/protobuf/cmd/protoc-gen-go@v1.31.0
+	go install google.golang.org/grpc/cmd/protoc-gen-go-grpc@v1.3.0
+
+install-go-ci-dependencies:
```
Comment (franciscojavierarceo) [inline Makefile:550]: this is all commented out, do we need it?

File: Makefile [hunk 7]
```diff
@@ -536,3 +536,40 @@ build-helm-docs:
...
+	# go install golang.org/x/tools/cmd/goimports
+	# python -m pip install "pybindgen==0.22.1" "grpcio-tools>=1.56.2,<2" "mypy-protobuf>=3.1"
+
+build-go: compile-protos-go
```
Comment (franciscojavierarceo) [inline Makefile:554]: is this make command correct?

Comment (franciscojavierarceo) [inline Makefile:554]: the inline seems to be at odds with the rest of the Makefile format, could you adjust it?

File: Makefile [hunk 8]
```diff
@@ -536,3 +536,40 @@ build-helm-docs:
...
+install-feast-ci-locally:
+	pip install -e ".[ci]"
+
+test-go: compile-protos-go compile-protos-python install-feast-ci-locally
+	CGO_ENABLED=1 go test -coverprofile=coverage.out ./... && go tool cover -html=coverage.out -o coverage.html
+
+format-go:
+	gofmt -s -w go/
+
+lint-go: compile-protos-go
```
Comment (franciscojavierarceo) [inline Makefile:566]: same here

File: go/internal/feast/onlinestore/sqliteonlinestore.go [hunk 9]
```diff
@@ -41,6 +41,7 @@ func NewSqliteOnlineStore(project string, repoConfig *registry.RepoConfig, onlin
 			return nil, fmt.Errorf("cannot find convert sqlite path to string %s", db_path)
 		} else {
 			store.path = fmt.Sprintf("%s/%s", repoConfig.RepoPath, dbPathStr)
+
```
Comment (franciscojavierarceo) [inline go/internal/feast/onlinestore/sqliteonlinestore.go:44]: was this just a lint format change?

File: go/internal/feast/registry/registry.go [hunk 10]
```diff
@@ -35,24 +37,25 @@ type Registry struct {
...
 }
 
-func NewRegistry(registryConfig *RegistryConfig, repoPath string) (*Registry, error) {
+func NewRegistry(registryConfig *RegistryConfig, repoPath string, project string) (*Registry, error) {
 	registryStoreType := registryConfig.RegistryStoreType
 	registryPath := registryConfig.Path
 	r := &Registry{
-		cachedRegistryProtoTtl: time.Duration(registryConfig.CacheTtlSeconds),
+		project:                project,
+		cachedRegistryProtoTtl: time.Duration(registryConfig.CacheTtlSeconds) * time.Second,
```
Comment (franciscojavierarceo) [inline go/internal/feast/registry/registry.go:48]: This seems like a pretty large change in time proportional to the number of seconds, no? Was this on purpose?

File: go/internal/feast/transformation/transformation.go [hunk 11]
```diff
@@ -91,103 +99,6 @@ func ReleaseArrowContext(requestContextArrow map[string]arrow.Array) {
 	}
 }
 
-func CallTransformations(
```
Comment (franciscojavierarceo) [inline go/internal/feast/transformation/transformation.go:94]: rip

File: Makefile (hunk 6 above)
Comment (shuchu) [reply Makefile:550]: commented out as I'm not sure if this setting is good/ready for CI test.

File: go/internal/feast/registry/registry.go (hunk 10 above)
Comment (shuchu) [reply go/internal/feast/registry/registry.go:48]: This one may need @EXPEbdodla 's help to explain.

File: Makefile (hunk 8 above)
Comment (shuchu) [reply Makefile]: done

File: Makefile [hunk 12]
```diff
@@ -536,3 +536,40 @@ build-helm-docs:
...
+	# go install golang.org/x/tools/cmd/goimports
+	# python -m pip install "pybindgen==0.22.1" "grpcio-tools>=1.56.2,<2" "mypy-protobuf>=3.1"
+
+build-go: compile-protos-go
+	go build -o feast ./go/main.go
+
+install-feast-ci-locally:
+	pip install -e ".[ci]"
+
+test-go: compile-protos-go compile-protos-python install-feast-ci-locally
```
[the remark this answers is not in the document]
Comment (shuchu) [reply Makefile]: done

File: Makefile (hunk 7 above)
Comment (shuchu) [reply Makefile]: done

File: go/internal/feast/onlinestore/redisonlinestore.go [hunk 13]
```diff
@@ -119,31 +147,35 @@ func getRedisType(onlineStoreConfig map[string]interface{}) (redisType, error) {
...
 		if _, ok := featureViewIndices[featureViewName]; !ok {
 			featureViewIndices[featureViewName] = index
 			indicesFeatureView[index] = featureViewName
 			index += 1
 		}
 	}
+	return featureViewIndices, indicesFeatureView, index
+}
+
+func (r *RedisOnlineStore) buildHsetKeys(featureViewNames []string, featureNames []string, indicesFeatureView map[int]string, index int) ([]string, []string) {
```
[the remark this answers is not in the document]
Comment (shuchu) [reply go/internal/feast/onlinestore/redisonlinestore.go]: done

File: go/internal/feast/onlinestore/sqliteonlinestore.go (hunk 9 above)
Comment (shuchu) [reply go/internal/feast/onlinestore/sqliteonlinestore.go]: fixed.

**Review by franciscojavierarceo — APPROVED** [review]
🚀 🚀

**Pull request merged**